concurrent/src/main/java/io/airlift/concurrent/AsyncSemaphore.java

```diff
@@ -37,26 +42,70 @@
  * to do the bulk of its work asynchronously.
  */
 @ThreadSafe
-public class AsyncSemaphore<T>
+public class AsyncSemaphore<T, R>
 {
-    private final Queue<QueuedTask<T>> queuedTasks = new ConcurrentLinkedQueue<>();
+    private final Queue<QueuedTask<T, R>> queuedTasks = new ConcurrentLinkedQueue<>();
     private final AtomicInteger counter = new AtomicInteger();
     private final Runnable runNextTask = this::runNext;
     private final int maxPermits;
     private final Executor submitExecutor;
-    private final Function<T, ListenableFuture<?>> submitter;
+    private final Function<T, ListenableFuture<R>> submitter;
+
+    /**
+     * Process a list of tasks as a single unit
+     * (similar to {@link com.google.common.util.concurrent.Futures#allAsList(ListenableFuture[])})
+     * with limiting the number of tasks running in parallel.
+     * <p>
+     * This method may be useful for limiting the number of concurrent requests sent to a remote server when
+     * trying to load multiple related entities concurrently:
+     * <p>
+     * For example:
+     * <pre>{@code
+     * List<Integer> userIds = Lists.of(1, 2, 3);
+     * ListenableFuture<List<UserInfo>> future = processAll(ids, client::getUserInfoById, 2, executor);
+     * List<UserInfo> userInfos = future.get(...);
+     * }</pre>
+     *
+     * @param tasks tasks to process
+     * @param submitter task submitter
+     * @param maxPermits maximum number of tasks permitted to run in parallel
```

Comment:
maxPermits => maxConcurrency (makes more sense when you are looking at processing in general, and not looking at an explicit semaphore)